qiskit/circuit/controlflow/if_else.py

```diff
@@ -0,0 +1,152 @@
+from .control_flow import ControlFlowOp
+
+
+class IfElseOp(ControlFlowOp):
+    """A circuit operation which executes a program (``true_body``) if a
+    provided condition (``condition``) evaluates to true, and
+    optionally evaluates another program (``false_body``) otherwise.
+
+    Parameters:
+        condition: A condition to be evaluated at circuit runtime which,
+            if true, will trigger the evaluation of ``true_body``. Can be
+            specified as either a tuple of a ``ClassicalRegister`` to be
+            tested for equality with a given ``int``, or as a tuple of a
+            ``Clbit`` to be compared to either a ``bool`` or an ``int``.
+        true_body: A program to be executed if ``condition`` evaluates
+            to true.
+        false_body: A optional program to be executed if ``condition``
+            evaluates to false.
+        label: An optional label for identifying the instruction.
+
+    If provided, ``false_body`` must be of the same ``num_qubits`` and
+    ``num_clbits`` as ``true_body``.
+
+    The classical bits used in ``condition`` must be a subset of those attached
+    to the circuit on which this ``IfElseOp`` will be appended.
+
+    **Circuit symbol:**
+
+    .. parsed-literal::
+
+             ┌───────────┐
+        q_0: ┤0          ├
+             │           │
+        q_1: ┤1          ├
+             │  if_else  │
+        q_2: ┤2          ├
+             │           │
+        c_0: ╡0          ╞
+             └───────────┘
+
+    """
+
+    def __init__(
+        self,
+        condition: Union[
+            Tuple[ClassicalRegister, int],
+            Tuple[Clbit, int],
+            Tuple[Clbit, bool],
+        ],
+        true_body: QuantumCircuit,
+        false_body: Optional[QuantumCircuit] = None,
+        label: Optional[str] = None,
+    ):
+        # Type checking generally left to @params.setter, but required here for
+        # finding num_qubits and num_clbits.
+        if not isinstance(true_body, QuantumCircuit):
+            raise CircuitError(
+                "IfElseOp expects a true_body parameter "
+                f"of type QuantumCircuit, but received {type(true_body)}."
+            )
+
+        num_qubits = true_body.num_qubits
+        num_clbits = true_body.num_clbits
+
+        super().__init__("if_else", num_qubits, num_clbits, [true_body, false_body], label=label)
+
+        try:
+            lhs, rhs = condition
+        except TypeError as err:
+            raise CircuitError(
+                "IfElseOp expects a condition argument as either a "
+                "Tuple[ClassicalRegister, int], a Tuple[Clbit, bool] or "
+                f"a Tuple[Clbit, int], but received {condition} of type "
+                f"{type(condition)}."
+            ) from err
+
+        if not (
+            (isinstance(lhs, ClassicalRegister) and isinstance(rhs, int))
+            or (isinstance(lhs, Clbit) and isinstance(rhs, (int, bool)))
+        ):
```

Comment:
I left this in place for sake of readability (I'd not assume most folks know `bool` is a subclass of `int`, I didn't :) ), and it should be negligible in performance cost. If you feel strongly, I'd also be happy removing it.

Reply:
I'm happy enough either way.  (I think in my circuit-builders, I've written the condition type hints in shortened form, though - I may need to change them.)  Aside from the allocation of one additional Boolean, the performance of your form should always be identical in the good path (i.e. performs the same number of `__instancecheck__` calls) because `bool.__instancecheck__` will never be called.  In the failure path there's one more call, but we don't care about ludicrously small performance improvements there anyway.